core/reactor_common.c

```diff
@@ -1534,6 +1534,24 @@ lf_token_t* _lf_set_new_array_impl(lf_token_t* token, size_t length, int num_des
     return new_token;
 }
 
+/**
+ * Check the deadline of the currently executing reaction against the
+ * current physical time. If the deadline has passed, invoke the deadline
+ * handler and return true. Otherwise, return false.
+ * 
+ * @param self The self struct of the reactor.
+ * @return true if the specified deadline has passed.
+ * @return false if the deadline has not passed yet.
+ */
+bool _lf_check_deadline(self_base_t* self) {
+	reaction_t* reaction = self->executing_reaction;
+	if (get_physical_time() > get_logical_time() + reaction->deadline) {
+		reaction->deadline_violation_handler(self);
```

Comment:
@edwardalee The deadline violation handler function is called here. Is this semantically okay?

Because of the way `check_deadline()` is supposed to work, the deadline handler could be called multiple times (if the caller doesn't terminate the reaction (e.g., by using `return`), upon `check_deadline()` returning true, and calls `check_deadline()` again later). What if the deadline handler had some side effect, including changing a state variable and setting to an output port?
This semantic seems a bit strange to me. Depending on the body of the reaction, the deadline violation handler could be called multiple times within the same tag. Even stranger, the number of times the deadline violation handler is called depends on many external factors, like the load on the system.

Reply:
Okay, got it.

I think the way the API is documented on the website might be important to make sure programmers fully understand the ramifications of using `check_deadline`.

Reply:
I just merged the PR as I believe we reached an agreement. Please let me know if there's any further suggestion, I'll be happy to address that.